kevm-pyk/src/kevm_pyk/foundry.py

```diff
@@ -346,69 +346,93 @@ def foundry_prove(
     if unfound_tests:
         raise ValueError(f'Test identifiers not found: {unfound_tests}')
 
-    ag_proofs: dict[str, AGProof] = {}
-    for test in tests:
-        if AGProof.proof_exists(test, ag_proofs_dir) and not reinit:
-            ag_proof = AGProof.read_proof(test, ag_proofs_dir)
-        else:
-            _LOGGER.info(f'Initializing KCFG for test: {test}')
-            contract_name, method_name = test.split('.')
-            contract = foundry.contracts[contract_name]
-            method = [m for m in contract.methods if m.name == method_name][0]
-            empty_config = foundry.kevm.definition.empty_config(GENERATED_TOP_CELL)
-            kcfg = _method_to_cfg(empty_config, contract, method)
-
-            _LOGGER.info(f'Expanding macros in initial state for test: {test}')
-            init_term = kcfg.get_unique_init().cterm.kast
-            init_term = KDefinition__expand_macros(foundry.kevm.definition, init_term)
-            init_cterm = CTerm.from_kast(init_term)
-
-            _LOGGER.info(f'Expanding macros in target state for test: {test}')
-            target_term = kcfg.get_unique_target().cterm.kast
-            target_term = KDefinition__expand_macros(foundry.kevm.definition, target_term)
-            target_cterm = CTerm.from_kast(target_term)
-            kcfg.replace_node(kcfg.get_unique_target().id, target_cterm)
-
-            _LOGGER.info(f'Starting KCFGExplore for test: {test}')
-            with KCFGExplore(
-                foundry.kevm,
-                id=test,
-                bug_report=br,
-                kore_rpc_command=kore_rpc_command,
-                smt_timeout=smt_timeout,
-                smt_retry_limit=smt_retry_limit,
-            ) as kcfg_explore:
-                _LOGGER.info(f'Computing definedness constraint for test: {test}')
-                init_cterm = kcfg_explore.cterm_assume_defined(init_cterm)
-                kcfg.replace_node(kcfg.get_unique_init().id, init_cterm)
-
-                if simplify_init:
-                    _LOGGER.info(f'Simplifying KCFG for test: {test}')
-                    kcfg_explore.simplify(kcfg)
-
-            ag_proof = AGProof(test, kcfg, proof_dir=ag_proofs_dir)
+    setup_methods: dict[str, str] = {}
+    contracts = unique({test.split('.')[0] for test in tests})
+    for contract_name in contracts:
+        contract = foundry.contracts[contract_name]
+        method = contract.method_by_name('setUp')
+        if method is not None:
+            setup_methods[contract.name] = f'{contract.name}.{method.name}'
+
+    def run_cfg_group(tests: list[str]) -> dict[str, bool]:
+        ag_proofs: dict[str, AGProof] = {}
+        for test in tests:
+            if AGProof.proof_exists(test, ag_proofs_dir) and not reinit:
+                ag_proof = AGProof.read_proof(test, ag_proofs_dir)
+                assert type(ag_proof) is AGProof
```

Comment:
```suggestion
                ag_proof = AGProof.read_proof(test, ag_proofs_dir)
```

We don't need this assertion anymore.